### DISCUSSION TITLE
Single quote on component with wire:click and LivewireUI Modal

### DISCUSSION THREAD

**Discussion by mattkibbler** [livewire/livewire#4115] [1. Help]
Hi there,

I'm not sure if this is a Livewire or Laravel thing, but I was trying to user the [LivewireUI Modal](https://github.com/livewire-ui/modal) package to create and open a modal. 

Following the LivewireUI Modal documentation, in order to mount the modal with data you need to use single quotes for the `wire:click` attribute like so:

```html
<!-- Outside of any Livewire component -->
<button onclick='Livewire.emit("openModal", "edit-user", {{ json_encode(["user" => $user->id]) }})'>Edit User</button>

<!-- Inside existing Livewire component -->
<button wire:click='$emit("openModal", "edit-user", {{ json_encode(["user" => $user->id]) }})'>Edit User</button>

<!-- If you use a different primaryKey (e.g. email), adjust accordingly -->
<button wire:click='$emit("openModal", "edit-user", {{ json_encode(["user" => $user->email]) }})'>Edit User</button>

<!-- Example of passing multiple parameters -->
<button wire:click='$emit("openModal", "edit-user", {{ json_encode([$user->id, $isAdmin]) }})'>Edit User</button>

```

This works fine on a plain HTML button, like so:

```html
<button wire:click='$emit("openModal", "test-modal", @json(["user" => 1]))' class="bg-gray-400 px-4 py-2 rounded-md text-white">Open test modal</button>
```

... but when using a component like this:

```html
<x-button wire:click='$emit("openModal", "test-modal", @json(["user" => 1]))'>Open test modal</x-button>
```

... it causes problems. The button doesn't render properly, instead it looks like this:

![image](https://user-images.githubusercontent.com/3206141/139242429-37dc9371-92d4-4bc0-b3dc-51a19226e555.png)

...and when clicking the button component I get [this exception](https://flareapp.io/share/DPyk2Lb5#F49).

`
Livewire\Exceptions\MethodNotFoundException
Unable to call component method. Public method [$emit(\] not found on component: [test-modal-container]
`

I've set up an example repository [here](https://github.com/mattkibbler/livewire-modal-bug) to demonstrate this.

Please let me know if I'm doing something wrong or have misunderstood how I'm supposed to achieve this.

Cheers!

**Comment by joshhanley** [comment]
@mattkibbler the issue is `@json` doesn't output valid JSON that is safe to be echo'ed out in HTML like that. 

For that reason we have an `@js` directive that was added in Livewire 2.7. You can checkout the docs for it here under "Using the `@js` directive" https://laravel-livewire.com/docs/2.x/alpine-js

Hope this helps!

**Reply by mattkibbler** [reply]
Thanks for the reply, @joshhanley.

`@js` doesn't do it because this isn't for AlpineJS, I just need to get JSON onto the page, I see now that I should just be using `json_encode`, the problem seems to be that if you use single quotes for the `wire:click` attribute like this:

```html
<x-button wire:click='$emit("openModal", "test-modal", {{json_encode(["user" => 1])}})'>Open test modal</x-button>
```

... then all double quotes inside that attribute get escaped and you end up with HTML like this:

```html
<button type="button" wire:click="$emit(\"openModal\", \"test-modal\", {&quot;user&quot;:1})">
    Open test modal
</button>
```
... which gives you a `Public method [$emit(\] not found on component:` error when clicking the button.

I've gotten this working by using single quotes in the `json_encode` like this:

```html
<x-button wire:click="$emit('openModal', 'test-modal', {{json_encode(['user' => 1])}})">Open test modal</x-button>
```
...which produces this HTML:
```html
<button type="button" wire:click="$emit('openModal', 'test-modal', {&quot;user&quot;:1})">
    Open test modal
</button>
```

I guess this might be a problem with the documentation in the modal package so I'll move any further discussion over there.

Thanks for your time.

**Comment by joshhanley** [comment]
@mattkibbler actually you can use `@js` for Alpine and plain javascript, you can also find it on this page https://laravel-livewire.com/docs/2.x/inline-scripts

The `@js` directive is for this exact scenario, which is getting data from PHP into Javascript in correctly formatted JSON but that is HTML safe (hence the escaped quotes), but allows you to use double quotes for `wire:click="..."`.

You will find the output is the same when you use single quotes with `json_encode` as described in the package.

The reason is the browser is (stupidly) changing the quotes when it gets presented in the elements tab in dev tools. But if you actually "view source" on the page it will be correct, hence why the code will actually run as expected.

For example, this code:

```blade
 <button wire:click='$emit("openModal", "delete-user", {{ json_encode(["user" => 1]) }})'>Delete User</button>
 ```
 
 Looks like this in dev tools
 
![image](https://user-images.githubusercontent.com/882837/139602866-8f42c6eb-423d-437a-b330-8b105fef6682.png)

But when you view source

![image](https://user-images.githubusercontent.com/882837/139602875-1c404eac-cbf1-4853-9e0e-c26245e55319.png)

You can see it is actually correct.

This is just our browsers trying to make things fancy in devtools, but actually being more of a pain (in my opinion) as it's presented wrong.

The code is actually correct and will run happily.

@PhiloNL FYI if you weren't aware of this 🙂

Hope this helps!
